pkg/election/election.go

```diff
@@ -171,46 +223,94 @@ func (e *Election) campaignLoop(ctx context.Context, session *concurrency.Sessio
 		}
 
 		// try to campaign
-		elec := concurrency.NewElection(session, e.key)
-		err = elec.Campaign(ctx, e.id)
-		if err != nil {
-			// err may be ctx.Err(), but this can be handled in `case <-ctx.Done()`
-			e.l.Warn("fail to campaign", zap.Error(err))
-			continue
+		if !e.isLeader.Get() {
+			elec := concurrency.NewElection(session, e.key)
+			val, _ := json.Marshal(e.meta)
+			err = elec.Campaign(ctx, string(val))
+			if err != nil {
+				// err may be ctx.Err(), but this can be handled in `case <-ctx.Done()`
+				e.l.Warn("fail to campaign", zap.Error(err))
+				continue
+			}
+
+			// compare with the current leader
+			_, leaderMeta, err := getLeaderInfo(ctx, elec)
+			if err != nil {
+				// err may be ctx.Err(), but this can be handled in `case <-ctx.Done()`
+				e.l.Warn("fail to get leader ID", zap.Error(err))
+				continue
+			}
+			// no leader available, leader key maybe delete, continue campaign
+			if leaderMeta == nil {
+				continue
+			}
+
+			if leaderMeta.ID != e.meta.ID {
+				e.l.Info("current member is not the leader", zap.String("current member", e.meta.ID), zap.String("leader", leaderMeta.ID))
+				// leader change, must update lead info and gRPC client to leader
+				if *leaderMeta != e.leaderMeta.Load() {
+					for i := 0; i < 3; i++ {
+						err = e.updateLeader(leaderMeta)
+						if err == nil {
+							break
+						}
+						log.L().Error("update leader failed", zap.Stringer("leader", leaderMeta))
+						time.Sleep(200 * time.Millisecond)
+					}
+				}
+
+				continue
+			}
+
+			e.toBeLeader() // become the leader now
 		}
 
-		// compare with the current leader
-		leaderKey, leaderID, err := getLeaderInfo(ctx, elec)
+		e.watchLeader(ctx, session, e.key)
+		_, leaderMeta, err := e.LeaderInfo(ctx)
 		if err != nil {
-			// err may be ctx.Err(), but this can be handled in `case <-ctx.Done()`
-			e.l.Warn("fail to get leader ID", zap.Error(err))
-			continue
-		}
-		if leaderID != e.id {
-			e.l.Info("current member is not the leader", zap.String("current member", e.id), zap.String("leader", leaderID))
-			continue
+			log.L().Warn("seems no leader is available, ")
+		} else {
+			if err := e.retireLeader(leaderMeta); err != nil {
+				e.ech <- err
+			}
 		}
-
-		e.toBeLeader() // become the leader now
-		e.watchLeader(ctx, session, leaderKey)
-		e.retireLeader() // need to re-campaign
 	}
 }
 
+func (e *Election) tryCampaignLeader(ctx context.Context, leaseID clientv3.LeaseID) error {
+	value, _ := json.Marshal(&e.meta)
+	_, err := kv.NewSlowLogTxn(e.cli).
+		If(clientv3.Compare(clientv3.CreateRevision(e.key), "=", 0)).
+		Then(clientv3.OpPut(e.key, string(value), clientv3.WithLease(leaseID))).
+		Commit()
+	return errors.WithStack(err)
+}
+
 func (e *Election) toBeLeader() {
 	e.isLeader.Set(true)
+	e.leaderMeta.Store(e.meta)
```

Comment:
Is it needed to combine `isLeader` into `leaderMeta`?

Reply:
yes, I remove field isLeader and use leaderMeta to judge whether current node is leader so we can gain high consistency with only one atomic operation.